Fixes 4-channel limit bug & improves preferences

petebankhead opened the pull request.

petebankhead committed:
Fix for multichannel images with > 4 channels

Fixed a bug in merging channels that prevented images with more than 4
channels being opened.

Also improved preferences to give more fine-grained control over
options related to performance and memoization.

petebankhead committed:
Updated readme with description of the preferences

petebankhead commented:
Thanks Oli, in the meantime I have also been looking into this a bit since having a few very helpful discussions with some of the OME team - and started testing an alternative way of estimating downsample values that will _hopefully_ behave more consistently for various file types and depend less on VSI as a special case.

I'm currently travelling, but hope to have a chance to look at this properly next week.  I'll work through the changes we've both been making to come up with a proposed new release version.

lacan commented:
Hi Pete, thanks for the heads up. Happy travels.

The frustrating thing is that for the magnification, the following code works just fine in Fiji: 
there is always one instrument index, and meta finds it for the right series, but in QuPath this leads to a NullPointer Exception...
Tested on the same file, in Fiji and QuPath...

```import loci.formats.ImageReader
import loci.formats.meta.IMetadata
import loci.common.services.ServiceFactory
import loci.formats.IFormatReader
import loci.formats.services.OMEXMLService


imagePath = "Z:/biop/Personel/Olivier/Projects/QuPath/Development/demo-project/Images/VSI/Bianca_1.vsi"


factory = new ServiceFactory()
service = factory.getInstance(OMEXMLService.class)
meta = service.createOMEXMLMetadata()

reader = new ImageReader()
reader.setFlattenedResolutions(false)
reader.setMetadataStore(meta)
reader.setId(imagePath)

seriesIdx = 2;

reader.setSeries(seriesIdx)

printf("reader series: %d, meta series: %d\n\n", reader.getSeriesCount(), meta.getImageCount())
printf("Instrument Count: %d\n", meta.getInstrumentCount())


instrumentRef = meta.getImageInstrumentRef(seriesIdx)
// The line above causes an error in QuPath...
printf("InstrumentRef: %s\n\n", instrumentRef)
```

petebankhead commented:
Odd - I'll be on the lookout for that.
But there's no QuPath-specific code there... right?  Can it be something to do with the exact Bio-Formats version / jars being used...?

lacan commented:
It could have something to do with the reader being multi threaded or synchronized, but seeing as you have a primary reader, which is the one that holds the metadata, this should not really be a problem.
I am a bit confounded, I must admit.

petebankhead commented:
If you've time, you could have a look at [v0.0.7](https://github.com/qupath/qupath-bioformats-extension/tree/v0.0.7) - it is a more thorough revision, which I hope addresses the issues (and others).

In particular, I tried to request instruments/objective magnifications in a way that I think makes sense and doesn't presuppose too much... but if you see any problems please let me know.
